Allow for caching of compaction scores

martinsumner opened the pull request:
Potentially reduce the overheads of scoring each file on every run.

The change also alters the default thresholds for compaction to favour longer runs (which will tend towards greater storage efficiency).

martinsumner committed:
Allow for caching of compaction scores

Potentially reduce the overheads of scoring each file on every run.

The change also alters the default thresholds for compaction to favour longer runs (which will tend towards greater storage efficiency).

martinsumner committed:
Clarify how the new option will be controlled in Riak

martinsumner committed:
Don't hide option

martinsumner committed:
Set max limit of 24 hours on cached score

martinsumner committed:
Improve the quality of score

Move the average towards the current score if not scoring each run.   Score from more keys to get a better score (as overheads of scoring are now better sorted by setting score_onein rather than by reducing the sample size).

martinsumner committed:
Standardise retention decision

Use the same function to decide for both scoring and compaction - and avoid the situation where somethig is scored for cmpaction, but doesnt change (which was the case previously with tombstones that were still in the ledger).

ThomasArts commented on priv/leveled.schema:
```diff
@@ -143,7 +149,7 @@
 %% then it is a candidate (e.g. in default case if 50% of space would be
```
```suggestion
%% then it is a candidate (e.g. in default case if 30% of space would be
```

martinsumner replied:
Amended - note that the percentage is the inverse of the space released

ThomasArts commented on src/leveled_cdb.erl:
```diff
@@ -424,6 +432,20 @@ cdb_isrolling(Pid) ->
+-spec cdb_getcachedscore(pid(), erlang:timestamp()) -> undefined|float().
+%% @doc
+%% Return the cached score for a CDB file
+cdb_getcachedscore(Pid, Now) ->
```
why not computing the timestamp inside the call? A bit weird to have it passed in through the interface, but probably that's for testing...

```
cdb_getcachedscore(Pid) ->
    gen_fsm:sync_send_all_state_event(Pid, {get_cachedscore, erlang:timestamp()}, infinity).
```

ThomasArts commented on src/leveled_cdb.erl:
```diff
@@ -2396,6 +2435,16 @@ get_keys_byposition_manykeys_test_to() ->
+    ?assertMatch(80.0, cdb_getcachedscore(P2, os:timestamp())),
+    timer:sleep(1000),
+    {NowMega, NowSecs, _} = Now = os:timestamp(),
+    ?assertMatch(80.0, cdb_getcachedscore(P2, Now)),
```
here you can fake "os:timestamp" by remembering the old and add 1000 to its seconds.
No need to sleep for a second and make testing slower.
Or, what else is there purpose of the API with a timestamp?

ThomasArts commented on src/leveled_iclerk.erl:
```diff
@@ -549,44 +584,31 @@ safely_log_filescore(PositionList, FN, Score, SW) ->
     leveled_log:log_timer("IC004", [Score, AvgJump, FN], SW).
 
 -spec size_comparison_score(list(key_size() | corrupted_test_key_size()),
```
It's a bit confusing that there is a `leveled_inker:filterfun()` type and a `leveled_cdb:filter_fun()`. Sooner or later this is going to be messed up.

ThomasArts commented on src/leveled_iclerk.erl:
```diff
@@ -549,44 +584,31 @@ safely_log_filescore(PositionList, FN, Score, SW) ->
-                            end
+                            {ActSize + Size - ?CRC_SIZE, RplSize};
+                        convert ->
+                            {ActSize, RplSize  + Size - ?CRC_SIZE};
```
```suggestion
                            {ActSize, RplSize + Size - ?CRC_SIZE};
```

convert and false are actually the same...

ThomasArts commented on src/leveled_iclerk.erl:
```diff
@@ -549,44 +584,31 @@ safely_log_filescore(PositionList, FN, Score, SW) ->
-                                    {ActSize, RplSize + Size - ?CRC_SIZE}
-                            end;
+                    ToRetain =
+                        to_retain({SQN, Type, PK},
```
this code gets computationally much heavier using `to_retail` instead of only `is_full_journalentry`....

martinsumner replied:
This is true, but the additional check is necessary to resolve the issue, and the extra load is compensated for by the act of caching

ThomasArts commented on src/leveled_iclerk.erl:
```diff
@@ -549,44 +584,31 @@ safely_log_filescore(PositionList, FN, Score, SW) ->
                         true ->
-                            Check = FilterFun(FilterServer, PK, SQN),
-                            case {Check, SQN > MaxSQN} of
-                                {current, _} ->
```
the old code can take this path without strategy being "retain" but new code cannot.

ThomasArts reviewed (changes requested):
Need to discuss the differences in behaviour for retain once more

martinsumner committed:
Promote cache when scanning

When scanning over a leveled store with a helper (e.g. segment filter and last modified date range), applying the filter will speed up the query when the block index cache is available to get_slots.

If it is not available, previously the leveled_sst did not then promote the cache after it had accessed the underlying blocks.

Now the code does this, and also when the cache has all been added, it extracts the largest last modified date so that sst files older than the passed in date can be immediately dismissed

martinsumner committed:
Switch to checking on get_kvrange

In production scale testing, placing te check_modified call on get_kvrange not get_slots made the performance difference.

It should help in get_lots as well, but unable to reliably get coverage in tests with this.  So for now, will leave off until a proper test can be constructed which demonstrates any benefits.

martinsumner committed:
Update src/leveled_iclerk.erl

Co-authored-by: Thomas Arts <thomas.arts@quviq.com>
